### DISCUSSION TITLE
ref to question #41- looking into using the SAMD21 XIAO with arduino audio-tools

### DISCUSSION THREAD

**Discussion by SnazzyBlanket** [pschatzmann/arduino-audio-tools#1439] [Q&A] [answered]
Based on what I have seen it looks like the chart shows that SAMD21 will work for I2s in and out (but nothing else?)
I would love to get this working as the advantage of a $1 chip with a built in DAC which runs arduino can not be understated. I think what you mentioned was I would need to add the adafruit i2s library (and then send the I2s output stream to that somehow?

OR could I not just use the callback stream and then send the output of that into the AnalogWrite (dacPin)?

**Comment by SnazzyBlanket** [comment]
THIS looks promising

Editing the Variant.h file seems to be the simplest solution.
An alternative is to select a board with I2S activated by default. (and the same processor)
I got my XIAO running with I2S by selecting an Adafruit Itsybitsy M0 as board.
Advantage is you don’t need to worry about board updates that remove your modifications to Variants.h
Disadvantage is that it’s a small puzzle to find out on whitch XIAO pins the Itsybitsy pins are.





from 

https://forum.seeedstudio.com/t/can-a-xiao-be-configured-for-i2s/253248/19

**Comment by SnazzyBlanket** [comment]
ok i can get i2s working BUT that does not allow me to route into the built in DAC. it seems to use SPI
https://github.com/hydronics2/SamdAudio/blob/master/SamdAudio.h

so I am thinking the best method may be to use a CALLBACK stream. within the callback stream send the output to AnalogWrite(pin of dac)

i am trying to understand how to do this and i found this callback example



`auto invert = [](uint8_t* data, size_t bytes) {
  size_t sample_count = bytes / sizeof(int16_t);
  int16_t* data16 = (int16_t*)data;
  for (int j = 0; j < sample_count; j++) {
    data16[j] = -data16[j];
  }
  return bytes;
};`


so could i replace the code inside of invert-- with analogWrite(dacpin)=data16[i] 

obviously using a counter to step through the data

**Comment by pschatzmann** [comment, marked as answer]
If you want to use the internal DAC, things get a little bit more complicated because you need to output the data at the correct rate.  I guess the easiest way forward would be if you extend my library and [provide an timer implementation](https://github.com/pschatzmann/arduino-audio-tools/tree/main/src/AudioTimer) for SAMD. You can try to port the one provided by your mentioned library.

This would allow you to use my default arduino implementation provided by AnalogAudioArduino.h
